crates/pico/src/memo.rs

```diff
@@ -0,0 +1,188 @@
+                node.dependencies = dependencies;
+                node.time_calculated = time_calculated;
+                node.time_verified = current_epoch;
+            } else {
+                db.storage_mut().derived_nodes().insert(
+                    node_id,
+                    DerivedNode {
+                        time_verified: current_epoch,
+                        time_calculated,
+                        dependencies,
+                        inner_fn,
+                        value,
+                    },
+                );
+                state = MemoState::Computed;
+            }
+            time_calculated
+        } else {
+            let node = db
+                .storage_mut()
+                .derived_nodes()
+                .get_mut(&node_id)
+                .expect("node should exist. This is indicative of a bug in Pico.");
+            node.time_verified = current_epoch;
+            node.time_calculated
+        }
+    } else {
+        let (value, dependencies, time_calculated) =
+            call_inner_fn_and_collect_dependencies(db, node_id.param_id, inner_fn);
+        db.storage_mut().derived_nodes().insert(
+            node_id,
+            DerivedNode {
+                time_verified: current_epoch,
+                time_calculated,
+                dependencies,
+                inner_fn,
+                value,
+            },
+        );
+        state = MemoState::Computed;
+        time_calculated
+    };
+    register_dependency(
+        db,
+        NodeKind::Derived(node_id),
+        time_calculated,
+        current_epoch,
+    );
+    state
+}
+
+fn any_dependency_changed<Db: Database>(
+    db: &mut Db,
+    node_id: &DerivedNodeId,
+    current_epoch: Epoch,
+) -> bool {
+    let dependencies = db
+        .storage()
+        .derived_nodes()
+        .get(node_id)
+        .expect("node should exist. This is indicative of a bug in Pico.")
+        .dependencies
+        .iter()
+        .filter_map(|dep| {
+            if dep.time_verified_or_calculated != current_epoch {
+                Some((dep.node_to, dep.time_verified_or_calculated))
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<_>>();
+
+    dependencies
+        .into_iter()
+        .any(|(dep_node, time_verified)| match dep_node {
+            NodeKind::Source(key) => source_node_changed(db, &key, time_verified),
+            NodeKind::Derived(dep_node_id) => derived_node_changed(db, dep_node_id),
+        })
+}
+
+fn source_node_changed<Db: Database>(db: &Db, key: &Key, time_verified: Epoch) -> bool {
+    match db.storage().source_nodes().get(key) {
+        Some(source) => source.time_calculated > time_verified,
+        None => true,
+    }
+}
+
+fn derived_node_changed<Db: Database>(db: &mut Db, node_id: DerivedNodeId) -> bool {
+    let inner_fn = if let Some(node) = db.storage().derived_nodes().get(&node_id) {
+        node.inner_fn
+    } else {
+        return true;
```

Comment:
can you comment on what this means? e.g. it means that this is the first time we're calculating the derived node (maybe?), or the derived node has been GC'd, so we don't have the value so we can't be sure it hasn't changed.

Or, is it always a bug if we're missing the inner fn?

Reply:
node_id here is a dependency, so we can't be sure it's not GC'd.

Reply:
I see. Maybe it's clearer if the fn is named `derived_node_must_be_recalculated`?

Reply:
We are in the `any_dependency_changed` fn, we check what dependencies must be verified and then for every found dependency we check if `this` dependency is changed or not. I'm worried if `derived_node_must_be_recalculated` fn with arguments `db` and `node_id` sounds like its own node must be recalculated, but it rather return a `bool` that indicates that a parent node must be recalculated (we've already done recalculation of dependent node inside `derived_node_changed` fn)